File: 48-rotate-image/rotate-image.py
class Solution:
    def rotate(self, matrix: List[List[int]]) -> None:
        """
        Do not return anything, modify matrix in-place instead.
        """
        n = len(matrix)
        def transpose(matrix,n):
            for i in range(n):
                for j in range(i,n):
                    matrix[i][j], matrix[j][i] = matrix[j][i], matrix[i][j]
        
        def reverse(matrix,n):
            for i in range(n):
                k = n - 1
                j = 0
                while j < k:
                    matrix[j][i], matrix[k][i] = matrix[k][i], matrix[j][i]
                    j += 1
                    k -= 1
            
        for i in range(3):
            transpose(matrix,n)
            reverse(matrix,n)

        

File: 48-rotate-image/rotate-image.java
class Solution {
    public void rotate(int[][] matrix) {
        int n = matrix.length;
        for (int i = 0; i < 3; i++){
            transpose(matrix,n);
            reverse(matrix,n);
        }
    }
    static void transpose(int[][] matrix,int n){
        for (int i = 0; i < n; i++){
            for(int j = i; j < n; j++){
                int temp = matrix[i][j];
                matrix[i][j] = matrix[j][i];
                matrix[j][i] = temp;
            }
        }
    }

    static void reverse(int[][] matrix,int n){
        for (int i  = 0; i < n; i++){
            for (int j = 0, k = n - 1; j < k; j++ , k--){
                int temp = matrix[j][i];
                matrix[j][i] = matrix[k][i];
                matrix[k][i] = temp;
            }
        }
    }
}